File: services/graphql-ims/src/models/sales_order_items.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, dataloader::Loader};
use chrono::{DateTime, Utc};
use graphql_core::PostgresDataLoader;
use uuid::Uuid;

use crate::models::{products, sales_orders};

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, async_graphql::SimpleObject, sqlx::FromRow)]
#[graphql(name = "ImsSalesOrderItems", complex)]
pub struct Model {
    pub id: Uuid,
    #[graphql(skip)]
    pub sales_order_id: Uuid,
    #[graphql(skip)]
    pub product_id: Uuid,
    pub quantity_ordered: i32,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    async fn sales_order(&self, ctx: &Context<'_>) -> async_graphql::Result<sales_orders::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(sales_orders::PrimaryKey(self.sales_order_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to get sales order"))?)
    }
    async fn product(&self, ctx: &Context<'_>) -> async_graphql::Result<products::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(products::PrimaryKey(self.product_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to get product"))?)
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results = sqlx::query_as::<_, Self::Value>(
            "select * from ims.sales_order_items where id = ANY($1)",
        )
        .bind(&keys)
        .fetch_all(&self.pool)
        .await?
        .into_iter()
        .map(|model| (PrimaryKey(model.id), model))
        .collect::<_>();

        Ok(results)
    }
}

File: packages/graphql/src/schema/wms/sales_order_items/resolvers/SalesOrderItems.ts
import {
  OutboundShipmentItems,
  SalesOrders,
  WmsProducts,
} from "../../../../zod.schema";
import type { SalesOrderItemsResolvers } from "./../../../types.generated";
export const SalesOrderItems: SalesOrderItemsResolvers = {
  salesOrder: async (parent, _args, ctx) => {
    const result = await ctx.db
      .selectFrom("wms.salesOrders")
      .selectAll("wms.salesOrders")
      .innerJoin(
        "wms.salesOrderItems",
        "wms.salesOrderItems.salesOrderId",
        "wms.salesOrders.id"
      )
      .where("wms.salesOrderItems.id", "=", parent.id as string)
      .executeTakeFirst();

    return result as unknown as SalesOrders;
  },
  product: async (parent, _args, ctx) => {
    const result = await ctx.db
      .selectFrom("wms.products")
      .selectAll("wms.products")
      .innerJoin(
        "wms.salesOrderItems",
        "wms.salesOrderItems.productId",
        "wms.products.id"
      )
      .where("wms.salesOrderItems.id", "=", parent.id as string)
      .executeTakeFirst();

    return result as unknown as WmsProducts;
  },
  outboundShipmentItems: async (parent, _args, ctx) => {
    const results = await ctx.db
      .selectFrom("wms.outboundShipmentItems")
      .selectAll("wms.outboundShipmentItems")
      .where(
        "wms.outboundShipmentItems.salesOrderItemId",
        "=",
        parent.id as string
      )
      .execute();

    return results as unknown as OutboundShipmentItems[];
  },
};
